cpp/dolfinx/common/sort.h

```diff
@@ -6,161 +6,105 @@
-/// @tparam T The size of the bitset, which corresponds to the number of
-/// bits necessary to represent a set of integers. For example, N = 96
-/// for mapping three std::int32_t.
-/// @tparam BITS The number of bits to sort at a time
-/// @param[in] array The array to sort
-/// @param[in] perm FIXME
-template <typename T, int BITS = 16>
-void argsort_radix(std::span<const T> array, std::span<std::int32_t> perm)
-{
-  static_assert(std::is_integral_v<T>, "Integral required.");
-
-  if (array.size() <= 1)
-    return;
-
-  const auto [min, max] = std::ranges::minmax_element(array);
-  T range = *max - *min + 1;
+    // index type (if no projection is provided it holds I == T)
+    using I = std::remove_cvref_t<std::invoke_result_t<P, T>>;
 
-  // Sort N bits at a time
-  constexpr int bucket_size = 1 << BITS;
-  T mask = (T(1) << BITS) - 1;
-  std::int32_t mask_offset = 0;
+    if (range.size() <= 1)
+      return;
 
-  // Compute number of iterations, most significant digit (N bits) of
-  // maxvalue
-  int its = 0;
-  while (range)
-  {
-    range >>= BITS;
-    its++;
-  }
+    T max_value = proj(*std::ranges::max_element(range, std::less{}, proj));
 
-  // Adjacency list arrays for computing insertion position
-  std::array<std::int32_t, bucket_size> counter;
-  std::array<std::int32_t, bucket_size + 1> offset;
-
-  std::vector<std::int32_t> perm2(perm.size());
-  std::span<std::int32_t> current_perm = perm;
-  std::span<std::int32_t> next_perm = perm2;
-  for (int i = 0; i < its; i++)
-  {
-    // Zero counter
-    std::ranges::fill(counter, 0);
+    // Sort N bits at a time
+    constexpr int bucket_size = 1 << BITS;
+    T mask = (T(1) << BITS) - 1;
 
-    // Count number of elements per bucket
-    for (auto cp : current_perm)
+    // Compute number of iterations, most significant digit (N bits) of
+    // maxvalue
+    int its = 0;
+    while (max_value)
     {
-      T value = array[cp] - *min;
-      std::int32_t bucket = (value & mask) >> mask_offset;
-      counter[bucket]++;
+      max_value >>= BITS;
+      its++;
     }
 
-    // Prefix sum to get the inserting position
-    offset[0] = 0;
-    std::partial_sum(counter.begin(), counter.end(), std::next(offset.begin()));
+    // Adjacency list arrays for computing insertion position
+    std::array<I, bucket_size> counter;
+    std::array<I, bucket_size + 1> offset;
 
-    // Sort py permutation
-    for (auto cp : current_perm)
+    I mask_offset = 0;
+    std::vector<T> buffer(range.size());
+    std::span<T> current_perm = range;
+    std::span<T> next_perm = buffer;
+    for (int i = 0; i < its; i++)
     {
-      T value = array[cp] - *min;
-      std::int32_t bucket = (value & mask) >> mask_offset;
-      std::int32_t pos = offset[bucket + 1] - counter[bucket];
-      next_perm[pos] = cp;
-      counter[bucket]--;
+      // Zero counter array
+      std::ranges::fill(counter, 0);
+
+      // Count number of elements per bucket
+      for (const auto& c : current_perm)
```

Comment:
Could this just be a copy rather than a reference?

Reply:
Before these changes it was a copy. Since the value type `T` however now is not restricted to trivially copyable types any longer, we should stick to `const auto&` I think.